go-controller/pkg/node/egressfirewall_dns/egressfirewall_dns.go

```diff
@@ -0,0 +1,271 @@
+
+	dnsobject "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/dnsobject/v1"
+	dnsobjectapi "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/dnsobject/v1"
+
+	errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+)
+
+type EgressDNS struct {
+	// Protects pdMap/namespaces operations
+	lock sync.Mutex
+	// holds DNS entries globally
+	dns *util.DNS
+	// name of the node
+	nodeName string
+	// this map holds dnsNames to the dnsEntries
+	dnsEntries map[string]*dnsEntry
+	// allows to get object using the informer chace
+	wf factory.NodeWatchFactory
+	k  kube.Interface
+
+	// Report change when Add operation is done
+	added          chan dnsNamespace
+	removed        chan string
+	stopChan       chan struct{}
+	controllerStop <-chan struct{}
+}
+
+type dnsEntry struct {
+	// the current IP addresses the dnsName resolves to
+	dnsResolves []net.IP
+	namespaces  map[string]struct{} //the namespaces that this DNSName applies to
+}
+
+type dnsNamespace struct {
+	dnsName   string
+	namespace string
+}
+
+func NewEgressDNS(nodeName string, watchFactory factory.NodeWatchFactory, k kube.Interface, controllerStop <-chan struct{}) (*EgressDNS, error) {
+	dnsInfo, err := util.NewDNS("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+
+	egressDNS := &EgressDNS{
+		dns:        dnsInfo,
+		nodeName:   nodeName,
+		dnsEntries: make(map[string]*dnsEntry),
+		wf:         watchFactory,
+		k:          k,
+
+		added:          make(chan dnsNamespace, 1),
+		removed:        make(chan string, 1),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(dnsNames []string, namespace string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	for _, dnsName := range dnsNames {
+		if _, exists := e.dnsEntries[dnsName]; !exists {
+			e.dnsEntries[dnsName] = &dnsEntry{
+				namespaces: map[string]struct{}{
+					namespace: {},
+				},
+			}
+			e.signalAdded(dnsNamespace{dnsName: dnsName, namespace: namespace})
+		} else {
+			// only need to add the namespace to the struct
+			e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+		}
+	}
+	return nil
+
+}
+
+func (e *EgressDNS) updateEntryForName(dnsNamespace dnsNamespace) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	ips := e.dns.GetIPs(dnsNamespace.dnsName)
+	//TODO: if the old and new resolutions are the same should not update the dnsObject
+	e.dnsEntries[dnsNamespace.dnsName].dnsResolves = ips
+
+	//update the dnsObject
+	dnsObject, err := e.wf.GetDNSObject(e.nodeName)
+	existed := true
+	if errors.IsNotFound(err) {
+		// the object *MOST LIKELY* has not been created yet
+		dnsObject = &dnsobjectapi.DNSObject{
+			ObjectMeta: metav1.ObjectMeta{Name: e.nodeName},
+			Spec:       dnsobjectapi.DNSObjectSpec{},
+		}
+		existed = false
+	} else if err != nil {
+		return err
+	}
+	var ipStrings []string
```

Comment:
Does this support dual-stack as-is?

Reply:
my understanding is that it is I have not had the opportunity to personally test it